// botframework-config/lib/models/qnaMakerService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module botframework-config
 */
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
const url_1 = require("url");
const schema_1 = require("../schema");
const connectedService_1 = require("./connectedService");
/**
 * Defines a QnA Maker service connection.
 */
class QnaMakerService extends connectedService_1.ConnectedService {
    /**
     * Creates a new QnaMakerService instance.
     * @param source (Optional) JSON based service definition.
     */
    constructor(source = {}) {
        super(source, schema_1.ServiceTypes.QnA);
        this.hostname = new url_1.URL('/qnamaker', this.hostname).href;
    }
    // encrypt keys in service
    encrypt(secret, encryptString) {
        if (this.endpointKey && this.endpointKey.length > 0) {
            this.endpointKey = encryptString(this.endpointKey, secret);
        }
        if (this.subscriptionKey && this.subscriptionKey.length > 0) {
            this.subscriptionKey = encryptString(this.subscriptionKey, secret);
        }
    }
    // decrypt keys in service
    decrypt(secret, decryptString) {
        if (this.endpointKey && this.endpointKey.length > 0) {
            this.endpointKey = decryptString(this.endpointKey, secret);
        }
        if (this.subscriptionKey && this.subscriptionKey.length > 0) {
            this.subscriptionKey = decryptString(this.subscriptionKey, secret);
        }
    }
}
exports.QnaMakerService = QnaMakerService;
