// botframework-connector/lib/connectorApi/models/conversationsMappers.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var mappers_1 = require("../models/mappers");
exports.ConversationsResult = mappers_1.ConversationsResult;
exports.ConversationMembers = mappers_1.ConversationMembers;
exports.ChannelAccount = mappers_1.ChannelAccount;
exports.ErrorResponse = mappers_1.ErrorResponse;
exports.ErrorModel = mappers_1.ErrorModel;
exports.InnerHttpError = mappers_1.InnerHttpError;
exports.ConversationParameters = mappers_1.ConversationParameters;
exports.Activity = mappers_1.Activity;
exports.ConversationAccount = mappers_1.ConversationAccount;
exports.MessageReaction = mappers_1.MessageReaction;
exports.SuggestedActions = mappers_1.SuggestedActions;
exports.CardAction = mappers_1.CardAction;
exports.Attachment = mappers_1.Attachment;
exports.Entity = mappers_1.Entity;
exports.ConversationReference = mappers_1.ConversationReference;
exports.TextHighlight = mappers_1.TextHighlight;
exports.SemanticAction = mappers_1.SemanticAction;
exports.ConversationResourceResponse = mappers_1.ConversationResourceResponse;
exports.ResourceResponse = mappers_1.ResourceResponse;
exports.Transcript = mappers_1.Transcript;
exports.PagedMembersResult = mappers_1.PagedMembersResult;
exports.AttachmentData = mappers_1.AttachmentData;
